omod/src/main/java/org/openmrs/module/registrationapp/form/RegisterPatientFormBuilder.java

```diff
@@ -61,40 +63,83 @@
+	 * @return the form structure
+	 * @should flatten the widget config
+	 * @should combine sections of widget config given 'combineSections' property set to true
 	 * @since 1.15.0
-	 * 
 	 * Builds the navigable form structure for the specified app descriptor combining sections
 	 * if specified
+	 */
+	public static NavigableFormStructure buildFormStructure(AppDescriptor app, Boolean combineSections) {
+		return buildFormStructure(app, combineSections, null, null);
+	}
+
+	/**
+	 * Builds the navigable form structure for the specified app descriptor
 	 *
 	 * @param app the app descriptor
+	 * @return the form structure
+	 * @should flatten the widget config
+	 */
+	public static NavigableFormStructure buildFormStructure(AppDescriptor app, AppFrameworkService appFrameworkService,
+			AppContextModel appContextModel) {
+		return buildFormStructure(app, false, appFrameworkService, appContextModel);
+	}
+
+	/**
+	 * @param app             the app descriptor
 	 * @param combineSections if true, sections are combined into one demographics section
 	 * @return the form structure
-	 * @throws IOException
 	 * @should flatten the widget config
 	 * @should combine sections of widget config given 'combineSections' property set to true
+	 * @since 1.15.0
+	 * Builds the navigable form structure for the specified app descriptor combining sections
+	 * if specified
 	 */
-	public static NavigableFormStructure buildFormStructure(AppDescriptor app, Boolean combineSections) throws IOException {
+	public static NavigableFormStructure buildFormStructure(AppDescriptor app, Boolean combineSections,
+			AppFrameworkService appFrameworkService, AppContextModel appContextModel) {
 		NavigableFormStructure formStructure = new NavigableFormStructure();
 
 		// Get the ordered list of sections out of the configuration
 		Map<String, Section> configuredSections = new LinkedHashMap<String, Section>();
 		ArrayNode sections = (ArrayNode) app.getConfig().get("sections");
+
+		// should we attempt to evaluate whether section or question should be included?
+		boolean evaluateContext = appFrameworkService != null && appContextModel != null;
 		for (JsonNode i : sections) {
 			ObjectNode config = (ObjectNode) i;
 
 			ObjectMapper objectMapper = new ObjectMapper();
 			Section section = objectMapper.convertValue(config, Section.class);
 
+			if (evaluateContext && !appFrameworkService.checkRequireExpression(section, appContextModel)) {
+				continue;
+			}
+
 			if (section.getQuestions() != null) {
-				for (Question question : section.getQuestions()) {
+				for (Iterator<Question> it = section.getQuestions().iterator(); it.hasNext(); ) {
+					Question question = it.next();
+					if (evaluateContext && !appFrameworkService.checkRequireExpression(question, appContextModel)) {
+						try {
+							it.remove();
+						}
+						catch (UnsupportedOperationException ignored) {}
+						catch (IllegalStateException ignored) {}
```

Comment:
Is it best practice to catch and ignore these? Should these ever be called in this case?

Reply:
The reason I left them ignored is because they should never happen, but I suppose that's a good a reason to throw them if they do happen.